### PR TITLE
Vector optimization for adding AES encryption and decryption to the loongarch64 architecture

### PR DESCRIPTION
**Pull request by zhuchen1911**
##### Checklist

- [ ] documentation is added or updated
- [ ] tests are added or updated

### CODE REVIEW

**Review by paulidale — APPROVED** [review]

File: crypto/loongarch_arch.h [hunk 1]
```diff
@@ -0,0 +1,18 @@
...
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+#ifndef OSSL_CRYPTO_LOONGARCH_ARCH_H
+#define OSSL_CRYPTO_LOONGARCH_ARCH_H
+
+//extern int have_lasx;
```
Comment (t8m) [inline crypto/loongarch_arch.h:12]: Please use regular C style comments.

**Review by t8m — CHANGES_REQUESTED** [review]

File: include/crypto/aes_platform.h [hunk 2]
```diff
@@ -158,6 +158,13 @@ void gcm_ghash_v8(u64 Xi[2],const u128 Htable[16],const u8 *inp, size_t len);
 #  define AESNI_CBC_HMAC_SHA_CAPABLE (OPENSSL_ia32cap_P[1]&(1<<(57-32)))
 # endif
 
+#if defined(__loongarch__) || defined(__loongarch64)
+#include "loongarch_arch.h"
+# if defined(VPAES_ASM)
+#   define VPAES_CAPABLE  (OPENSSL_loongarchcap_P & LOONGARCH_CFG2_LSX)
+# endif
+#endif
```
Comment (t8m) [inline include/crypto/aes_platform.h:166]: Please make the preprocessor directive indentations right. I.e., add one more space in front of everything, except the `#   define`, and add two spaces in `#include'.

Comment (zhuchen1911) [reply include/crypto/aes_platform.h]: Thank you very much for your review. preprocessor directive indentations has been corrected.

**Review by t8m — APPROVED** [review]
@paulidale still OK?

**Comment by paulidale** [comment]
Still good.

**Comment by paulidale** [comment]
Merged, thanks.

**Comment by cheungxi** [comment]
merge to `openssl-3.0` branch ?

**Comment by paulidale** [comment]
No, our stability policy prevent performance enhancements on stable branches.

It would be a possibility for 3.1 however.

**Comment by xen0n** [comment]
This breaks with upstream toolchains that does NOT yet have the LSX/LASX support (in fact support has [just been merged about a week ago](https://sourceware.org/git/?p=binutils-gdb.git;a=commitdiff;h=75b2f521b101d974354f6ce9ed7c054d8b2e3b7a)!), we have to probe for assembler support and disable the asm if support is absent.

**Pull request closed without merging**